Improve downflow of information in `psbt::Error`

arturomf94 opened the pull request:
See third point in https://github.com/rust-bitcoin/rust-bitcoin/issues/837

Closes: https://github.com/rust-bitcoin/rust-bitcoin/issues/1589

apoelstra commented on bitcoin/src/psbt/error.rs:
```diff
@@ -80,6 +80,18 @@ pub enum Error {
+    /// Parsing error indicating invalid leaf version
+    InvalidLeafVersion,
+    /// Parsing error indicating a Taproot error
+    TaprootError(&'static str),
```
For `InvalidPublicKey` and `InvalidSignature` I think we should hold the `secp256k1::Error`.

Ideally, we'd hold the underlying error for the other variants too, though I understand that holding an `io::Error` would force us to remove some `#derive`s which might be controversial.

Also, after this PR are we able to get rid of the `ParseFailed` variant?

apoelstra replied:
Also `TaprootError` does not need the `Error` suffix.

apoelstra commented on bitcoin/src/psbt/serialize.rs:
```diff
@@ -409,21 +409,20 @@ impl Deserialize for TapTree {
-
             let leaf_version = LeafVersion::from_consensus(*version)
-                .map_err(|_| Error::ParseFailed("Leaf Version Error"))?;
+                .map_err(|_| Error::TaprootError("Leaf Version Error"))?;
```
Should this one be `InvalidLeafVersion` rather than `TaprootError`?

arturomf94 replied:
Yes, I'll fix this.

arturomf94 commented:
Force-pushed [99278a5](https://github.com/rust-bitcoin/rust-bitcoin/pull/1591/commits/99278a5643901e54e6f8a4cdf59dcdcb9edafff8) to address comments. Specifically:

* Replace `Taproot` with `InvalidLeafVersion` - [comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/1591#discussion_r1085593869).
* Remove `Error` suffix from `TaprootError` - [comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/1591#discussion_r1085594277)
* Split `InvalidPublicKey` into `InvalidPublicKey` and `InvalidSecp256k1PublicKey` and split `InvalidSignature` into `InvalidEcdsaSignature` and `InvalidSchnorrSignature`. All now hold the corresponding error - [comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/1591#discussion_r1085593211)

PS - I'm not sure what's up with the tests. They ran fine locally. I ran `./contrib/test.sh` and it worked OK. I also used both stable and `1.47.0` RVs and the tests seemed fine :thinking: Could it be GitHub, or am I missing something?

apoelstra commented:
Can you try `cargo test --verbose --features=no-std --no-default-features` locally?

arturomf94 commented:
> Can you try cargo test --verbose --features=no-std --no-default-features locally?

Ah - that was it. Pushed [68abfdb](https://github.com/rust-bitcoin/rust-bitcoin/pull/1591/commits/68abfdb0b9dab9c4141c99a6be191327760bcf95) with fix.

Thanks for that! :)

apoelstra commented:
This looks great! There's still some work to be done on these errors but this is a big improvement, and I think further changes may need to wait until crate smashing has settled out a bit.

apoelstra reviewed (approved):
ACK 68abfdb0b9dab9c4141c99a6be191327760bcf95
